File: src/core/compile-commands-generator.ts
/**
 * Compile commands generator for IDE integration.
 *
 * Generates compile_commands.json files for clangd-based IDEs (VSCode, CLion, etc.)
 * to enable code completion, navigation, and IntelliSense for Unreal Engine projects.
 *
 * @module core/compile-commands-generator
 */

import chalk from 'chalk';
import path from 'path';
import fs from 'fs-extra';
import { execa } from 'execa';
import { EngineResolver } from './engine-resolver';
import { ProjectPathResolver } from './project-path-resolver';
import { Logger } from '../utils/logger';
import { TargetResolver } from './target-resolver';
import { formatError } from '../utils/error';
import { DEFAULTS } from '../utils/constants';
import { resolveUnrealBuildToolPath } from '../utils/unreal-paths';

/** Options for generating compile commands database. */
export interface CompileCommandsGenerateOptions {
  /** Build target (Editor, Game, Client, Server) */
  target?: string;
  /** Build configuration (Debug, DebugGame, Development, Shipping, Test) */
  config?: string;
  /** Target platform (Win64, Win32, Linux, Mac) */
  platform?: string;
  /** Path to project directory or .uproject file */
  project?: string;
  /** Path to Unreal Engine installation */
  enginePath?: string;
  /** Whether to include plugin sources in compile commands */
  includePluginSources?: boolean;
  /** Whether to include engine sources in compile commands */
  includeEngineSources?: boolean;
  /** Whether to use engine includes in compile commands */
  useEngineIncludes?: boolean;
  /** Suppress all output */
  silent?: boolean;
}

/**
 * Generates compile_commands.json for IDE integration (VSCode, CLion, etc.).
 * Uses UnrealBuildTool to generate the clang database for code completion and IntelliSense.
 */
export class CompileCommandsGenerator {
  /**
   * Generates compile_commands.json for the specified project and options.
   * @param options - Generation options including target, config, platform, and paths
   * @returns Promise resolving to the path of the generated compile_commands.json file
   * @throws Error if project file not found, UBT not found, or generation fails
   */
  static async generate(options: CompileCommandsGenerateOptions): Promise<string> {
    const projectPath = await ProjectPathResolver.resolveOrThrow(options.project || process.cwd());
    const silent = options.silent || false;
    const logger = new Logger({ silent });

    if (!(await fs.pathExists(projectPath))) {
      throw new Error(`Project file not found: ${projectPath}`);
    }

    const enginePath = await EngineResolver.resolveEnginePath({
      projectPath,
      enginePath: options.enginePath,
    });

    const projectDir = path.dirname(projectPath);
    const target = options.target || DEFAULTS.BUILD_TARGET;
    const platform = options.platform || DEFAULTS.BUILD_PLATFORM;
    const config = options.config || DEFAULTS.BUILD_CONFIG;
    const targetName = await this.resolveTargetName(projectPath, target);

    const ubtPath = await resolveUnrealBuildToolPath(enginePath);

    const targetNames = targetName.split(' ').filter(Boolean);

    const args = ['-mode=GenerateClangDatabase', `-Project=${projectPath}`];

    for (const resolvedTarget of targetNames) {
      args.push(`-Target="${resolvedTarget} ${platform} ${config}"`);
    }

    if (options.includePluginSources) {
      args.push('-IncludePluginSources');
    }

    if (options.includeEngineSources) {
      args.push('-IncludeEngineSources');
    }

    if (options.useEngineIncludes) {
      args.push('-UseEngineIncludes');
    }

    logger.info(
      `Generating compile commands for: ${chalk.bold(targetNames.join(', '))} | ${chalk.bold(platform)} | ${chalk.bold(config)}`
    );
    logger.info(`Project: ${projectPath}`);
    logger.info(`Engine: ${enginePath}`);
    logger.divider();

    logger.debug(`Executing: ${ubtPath} ${args.join(' ')}`);

    const childProcess = execa(ubtPath, args, {
      stdio: 'pipe',
      cwd: path.dirname(ubtPath),
      reject: false,
    });

    if (childProcess.stdout) {
      childProcess.stdout.on('data', (data: Buffer) => {
        logger.info(data.toString().trim());
      });
    }

    if (childProcess.stderr) {
      childProcess.stderr.on('data', (data: Buffer) => {
        logger.error(data.toString().trim());
      });
    }

    const result = await childProcess;

    if (!result || result.exitCode !== 0) {
      throw new Error(
        `Generate compile commands failed with exit code ${result?.exitCode ?? 'unknown'}`
      );
    }

    const vscodeDir = path.join(projectDir, '.vscode');
    await fs.ensureDir(vscodeDir);

    const targetCompileCommandsPath = path.join(vscodeDir, 'compile_commands.json');
    const engineCompileCommandsPath = path.join(enginePath, 'compile_commands.json');

    if (await fs.pathExists(engineCompileCommandsPath)) {
      logger.debug(
        'Found compile_commands.json at engine directory, copying to project .vscode directory'
      );
      await fs.copy(engineCompileCommandsPath, targetCompileCommandsPath);
      await fs.remove(engineCompileCommandsPath);
    }

    if (!(await fs.pathExists(targetCompileCommandsPath))) {
      throw new Error(
        `compile_commands.json not found at expected location: ${targetCompileCommandsPath}`
      );
    }

    await this.updateVSCodeSettings(projectDir, logger);

    return targetCompileCommandsPath;
  }

  /**
   * Resolves a generic target name (e.g., 'Editor', 'Game') to a specific project target.
   * Falls back to a default target if resolution fails.
   * @param projectPath - Path to the .uproject file
   * @param target - Generic target name to resolve
   * @returns Promise resolving to the resolved target name or a default target
   */
  private static async resolveTargetName(projectPath: string, target: string): Promise<string> {
    const resolved = await TargetResolver.resolveTargetName(projectPath, target);
    // If resolution fails (returns undefined), fall back to the original target
    return resolved ?? target;
  }

  /**
   * Updates VSCode settings to configure clangd and C/C++ extension for Unreal Engine development.
   * Creates or merges settings.json in the .vscode directory to point to the compile commands database.
   * @param projectDir - Path to the project directory
   * @param logger - Logger instance for output
   * @returns Promise resolving when settings have been updated
   */
  private static async updateVSCodeSettings(projectDir: string, logger: Logger): Promise<void> {
    const vscodeDir = path.join(projectDir, '.vscode');
    await fs.ensureDir(vscodeDir);
    const settingsPath = path.join(vscodeDir, 'settings.json');

    const clangdConfig = {
      'clangd.arguments': [
        '--compile-commands-dir=${workspaceFolder}/.vscode',
        '--background-index',
        '--j=8',
        '--index-store-path=.clangd/index',
        '--pch-storage=disk',
        '--limit-results=200',
        '--header-insertion=iwyu',
      ],
    };

    const cppConfig = {
      'C_Cpp.default.compileCommands': '${workspaceFolder}/.vscode/compile_commands.json',
      'C_Cpp.intelliSenseEngine': 'disabled',
    };

    let settings: Record<string, unknown> = {};

    if (await fs.pathExists(settingsPath)) {
      try {
        const content = await fs.readFile(settingsPath, 'utf-8');
        settings = JSON.parse(content);
      } catch (error) {
        logger.debug(
          `Failed to parse existing settings.json, starting fresh: ${formatError(error)}`
        );
        settings = {};
      }
    }

    settings = { ...settings, ...clangdConfig, ...cppConfig };

    await fs.writeFile(settingsPath, JSON.stringify(settings, null, 2));
    logger.success(`Updated VSCode settings: ${settingsPath}`);
  }
}

File: src/core/compile_commands_generator.rs
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

use anyhow::{Context, Result};

use crate::utils::logger::Logger;
use crate::utils::unreal_paths::resolve_ubt_path;

use super::engine_resolver::EngineResolver;
use super::project_path_resolver::ProjectPathResolver;
use super::target_resolver::TargetResolver;

pub struct CompileCommandsGenerator;

impl CompileCommandsGenerator {
    pub fn generate(
        target: &str,
        config: &str,
        platform: &str,
        project: Option<&str>,
        engine_path: Option<&str>,
        include_plugin_sources: bool,
        include_engine_sources: bool,
        use_engine_includes: bool,
    ) -> Result<PathBuf> {
        let (project_path, engine) =
            EngineResolver::resolve_project_and_engine(project, engine_path)?;

        let project_dir = ProjectPathResolver::project_dir(&project_path);

        // Resolve target name
        let available = TargetResolver::find_available_targets(&project_path);
        let resolved_target = TargetResolver::resolve_from_list(target, &available);

        let ubt_path = resolve_ubt_path(&engine)?;

        let mut args = vec![
            "-mode=GenerateClangDatabase".to_string(),
            format!("-Project={}", project_path.display()),
            format!("-Target={resolved_target} {platform} {config}"),
        ];

        if include_plugin_sources {
            args.push("-IncludePluginSources".to_string());
        }
        if include_engine_sources {
            args.push("-IncludeEngineSources".to_string());
        }
        if use_engine_includes {
            args.push("-UseEngineIncludes".to_string());
        }

        Logger::info(&format!(
            "Generating compile commands for {resolved_target} ({platform} {config})"
        ));

        let output = Command::new(&ubt_path)
            .args(&args)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .output()
            .with_context(|| "Failed to run UnrealBuildTool")?;

        let stdout = String::from_utf8_lossy(&output.stdout);
        for line in stdout.lines() {
            Logger::debug(line);
        }

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!(
                "UBT exited with code {}:\n{}\n{stderr}",
                output.status.code().unwrap_or(-1),
                stdout.trim_end()
            );
        }

        // Relocate compile_commands.json
        let source_cc = engine.join("compile_commands.json");
        let dest_dir = project_dir.join(".vscode");
        let dest_cc = dest_dir.join("compile_commands.json");

        if source_cc.exists() {
            std::fs::create_dir_all(&dest_dir)?;
            std::fs::rename(&source_cc, &dest_cc)
                .or_else(|_| std::fs::copy(&source_cc, &dest_cc).map(|_| ()))?;

            // Update VSCode settings
            Self::update_vscode_settings(&dest_dir)?;
        } else if dest_cc.exists() {
            // Already in the right place
        } else {
            Logger::warning("compile_commands.json not found after generation");
        }

        Ok(dest_cc)
    }

    fn update_vscode_settings(vscode_dir: &Path) -> Result<()> {
        let settings_path = vscode_dir.join("settings.json");

        let mut settings: serde_json::Map<String, serde_json::Value> =
            if settings_path.exists() {
                let content = std::fs::read_to_string(&settings_path)?;
                serde_json::from_str(&content).unwrap_or_default()
            } else {
                serde_json::Map::new()
            };

        settings.insert(
            "C_Cpp.default.compileCommands".to_string(),
            serde_json::Value::String(
                "${workspaceFolder}/.vscode/compile_commands.json".to_string(),
            ),
        );
        settings.insert(
            "C_Cpp.intelliSenseEngine".to_string(),
            serde_json::Value::String("disabled".to_string()),
        );
        settings.insert(
            "clangd.arguments".to_string(),
            serde_json::Value::Array(vec![
                serde_json::Value::String(
                    "--compile-commands-dir=${workspaceFolder}/.vscode".to_string(),
                ),
                serde_json::Value::String("--background-index".to_string()),
                serde_json::Value::String("--j=8".to_string()),
                serde_json::Value::String("--index-store-path=.clangd/index".to_string()),
                serde_json::Value::String("--pch-storage=disk".to_string()),
                serde_json::Value::String("--limit-results=200".to_string()),
                serde_json::Value::String("--header-insertion=iwyu".to_string()),
            ]),
        );

        let json = serde_json::to_string_pretty(&settings)?;
        std::fs::write(&settings_path, json)?;

        Ok(())
    }
}
